### ISSUE TITLE
Plugins repo should pin its dependencies

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#108555]
We are experiencing a tree breakage today because Mockito deprecated an API and published it an hour or so ago.

google_maps_flutter_web's integration tests were using `^5.0.0` for mockito.

This caused analysis to fail after mockito was published, and is requiring some manual work arounds to avoid.

There is a fix forward at https://github.com/flutter/plugins/pull/6153, but we can make this more robust by pinning all deps for tests in the plugins repo. This still won't help for user facing deps though.

/cc @stuartmorgan @christopherfujino @Piinks

/cc @keyonghan @godofredoc @Hixie - this is a case where manual intervention that is not a revert ends up being needed to get the tree green again. I don't see how we can avoid this kind of situation given the cycles we've introduced in our infra (testing plugins on framework in this case).

**Comment by stuartmorgan** [comment]
> I don't see how we can avoid this kind of situation given the cycles we've introduced in our infra (testing plugins on framework in this case).

The cycle isn't really the problem here; this was just as bad for flutter/plugins. The cycle just extended the pain to flutter/flutter.

The combination of:
- dependencies we don't control
- that are unpinned
- with deprecations as errors

Leaves us open to out-of-band breakage, because deprecation doesn't require a major version change in semver. We either need to pin everything from outside the tree, or turn off deprecations as errors. The latter would be annoying for long-term code health, so we should definitely try to do the former.

**Comment by stuartmorgan** [comment]
I need to verify that dev_dependencies can be pinned without `pub publish` complaining. If it doesn't we can add tooling to have a list of dependencies that must be pinned (or more safely, a list of dependencies that are allowed to be unpinned).

If it does, we can do something more complex like make the analyze step locally edit everything to convert ranges to pins before starting the analysis.

**Comment by christopherfujino** [comment]
> We either need to pin everything from outside the tree, or turn off deprecations as errors. The latter would be annoying for long-term code health, so we should definitely try to do the former.

Agreed, this sounds best. I built a pub autoroller for the framework repo, would it be useful to extend this to flutter/plugins? https://github.com/flutter/flutter/issues/106371

**Comment by stuartmorgan** [comment]
Potentially, although we might be able to just use the new Dart Dependabot instead. I'd need to look into whether we can easily configure it to specific packages (since we don't want to constantly roll our inter-package deps).

**Comment by dnfield** [comment]
We would still be exposed to non-dev dependencies breaking us OOB like this though.

And deprecation breakages are not the only way to get broken.

**Comment by stuartmorgan** [comment]
> We would still be exposed to non-dev dependencies breaking us OOB like this though.

Yes, but we have very few of those that aren't repo-internal.

If we want complete coverage there we'll have to do CI-time munging, because we absolutely don't want to pin exact versions for all plugin clients.

**Comment by christopherfujino** [comment]
> Potentially, although we might be able to just use the new Dart Dependabot instead. I'd need to look into whether we can easily configure it to specific packages (since we don't want to constantly roll our inter-package deps).

Oh yeah, good idea.

**Comment by Hixie** [comment]
> combination of:
> 
> * dependencies we don't control
> * that are unpinned
> * with deprecations as errors
> 
> Leaves us open to out-of-band breakage, because deprecation doesn't require a major version change in semver.

Semver doesn't always help either, even if you ignore the deprecation issue. When we weren't pinning flutter/flutter's dependencies we often found that breaking changes leaked into "non-breaking" updates, even after the relevant package author had carefully thought about whether what they were doing was breaking or not. The reality is that all changes are potentially breaking. Semver only applies to the contract that you intend to export, but dependencies actually depend on everything, even the unintended contracts. Make an algorithm internally slightly faster with no change to anything else and suddenly a race condition manifests differently and a downstream package's tests flaky tests start manifesting failures where they previously were (by sheer chance) not failing. And of course sometimes people make breaking changes by mistake or forget to update the version appropriately.

**Comment by stuartmorgan** [comment]
Right; I meant that we're subject to breakage even within the context of the official contract. We're currently explicitly allowing picking up deprecations, and also explicitly making deprecations errors.

**Comment by stuartmorgan** [comment]
> If we want complete coverage there we'll have to do CI-time munging, because we absolutely don't want to pin exact versions for all plugin clients.

It occurred to me that the better way of getting complete pinning would be to just check in a lock file for every example app in the repo. We would need a bot that updated them all very frequently though, since otherwise we wouldn't be testing our federated plugins the way (new or `ugrade`ing) clients would experience them.

**Comment by godofredoc** [comment]
engine and flutter/flutter already create lock files but we don't have a way to updating them automatically. I tried to enable dependabot for java ecosystem but it doesn't generate lock files. Let me reach out to OSSF and GOSST on this topic, they may have a few ideas on how to handle java lock files.

The bonus points of generating lock files is that we will get vulnerability scanning for free.

**Comment by godofredoc** [comment]
Thanks for the clarification. Dependabot for the pub ecosystem is setup in a per directory. E.g. https://github.com/flutter/codelabs/blob/main/.github/dependabot.yaml.

This is an example of how we use it in cocoon:

* Dependabot configuration for app_dart folder in cocoon https://github.com/flutter/cocoon/blob/main/.github/dependabot.yml#L55
* Example PR https://github.com/flutter/cocoon/pull/2633/files

**Comment by godofredoc** [comment]
I believe there is a dependabot cli tool that may help run updates from a luci build with more control over the parameters and freq.

**Comment by stuartmorgan** [comment]
I'm going to downgrade this to P2; we've pinned various dev dependencies, and also switch flutter/flutter's version of `analyze` of flutter/packages to `downgrade` first which insulates it from OOB breakage. So while this isn't solved yet, we've reduced the impact in practice.

I was re-reading the discussion above about PRs and dependabot and such, and was thinking that we could probably make this much simpler since we don't need to isolate specific deps; we want everything to stay as current as possible, we just want to control it so we don't get OOB breakage. Given that, we could probably have a task that runs, say, once a day that just deletes and recreates all the lock files.

**Comment by stuartmorgan** [comment]
This just came up again: `web` 0.4.1 deprecated things we were using, which caused OOB breakage in both flutter/packages and flutter/flutter, and the ultimate result was that the deprecation was rolled back in 0.4.2 just for us. So that was pretty much the worst of all worlds, where we both disrupted the Flutter team *and* the maintainers of the package we depend on.

I'll put together a doc to more thoroughly explore options for addressing this.

**Comment by stuartmorgan** [comment]
> I'll put together a doc to more thoroughly explore options for addressing this.

I've deprioritized this due to the change in Flutter repo policy that means we no longer include deprecation in analysis, which was the source of almost all of the remaining problems around not pinning exact versions.

There's still value in things being fully repeatable, but we aren't actively experiencing issues around this at the moment.
